docs/redux.md

```diff
@@ -0,0 +1,1253 @@
+```
+
+## 3.2.22 **Use Plain JavaScript Objects for State**
+
+Prefer using plain JavaScript objects and arrays for your state tree.
+
+```tsx
+const initialState = {
+  users: [],
+  posts: [],
+};
+
+function myReducer(state = initialState, action) {
+  switch (action.type) {
+    case 'ADD_USER':
+      return {
+        ...state,
+        users: [...state.users, action.payload],
+      };
+    case 'ADD_POST':
+      return {
+        ...state,
+        posts: [...state.posts, action.payload],
+      };
+    default:
+      return state;
+  }
+}
+```
+
+## 3.3 **Priority C Rules: Recommended**
+
+## 3.3.1 **Write Action Types as domain/eventName**
+
+Use the "domain/eventName" convention for readability.
+
+```tsx
+const ADD_TODO = 'todos/addTodo';
+const INCREMENT = 'counter/increment';
+```
+
+## 3.3.2 **Write Actions Using the Flux Standard Action Convention**
+
+Prefer using FSA-formatted actions for consistency.
+
+```tsx
+const addTodo = (text) => ({
+  type: 'todos/addTodo',
+  payload: text,
+});
+
+const fetchTodosFailure = (error) => ({
+  type: 'todos/fetchTodosFailure',
+  payload: error,
+  error: true,
+});
+```
+
+## 3.3.3 **Use Action Creators**
+
+Text
+
+```tsx
+const addTodo = (text) => ({
+  type: 'ADD_TODO',
+  payload: text,
+});
+
+const increment = () => ({
+  type: 'INCREMENT',
+});
+```
+
+## 3.3.4 **Use RTK Query for Data Fetching**
+
+Use RTK Query as the default approach for data fetching and caching.
+
+```tsx
+import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
+
+const api = createApi({
+  reducerPath: 'api',
+  baseQuery: fetchBaseQuery({ baseUrl: '/api' }),
+  endpoints: (builder) => ({
+    getTodos: builder.query({
+      query: () => 'todos',
+    }),
+  }),
+});
+
+export const { useGetTodosQuery } = api;
+```
+
+## 3.3.5 **Use Thunks and Listeners for Other Async Logic**
+
+Use the Redux thunk middleware for imperative logic.
+
+```tsx
+// Thunk for fetching data
+const fetchData = () => {
+  return async (dispatch) => {
+    dispatch({ type: 'FETCH_START' });
+    try {
+      const response = await fetch('/api/data');
+      const data = await response.json();
+      dispatch({ type: 'FETCH_SUCCESS', payload: data });
+    } catch (error) {
+      dispatch({ type: 'FETCH_FAILURE', error });
+    }
+  };
+};
+```
+
+## 3.3.6 **Move Complex Logic Outside Components**
+
+Move complex synchronous or async logic outside components, usually into thunks.
+
+```tsx
+// Thunk for handling complex logic
+const complexLogic = () => {
+  return (dispatch, getState) => {
+    const state = getState();
+    // Perform complex logic here
+    dispatch({ type: 'COMPLEX_LOGIC_DONE' });
+  };
+};
+```
+
+## 3.3.7 **Use Selector Functions to Read from Store State**
+
+Use memoized selector functions for reading store state whenever possible.
+
+```tsx
+import { createSelector } from 'reselect';
+
+const selectTodos = (state) => state.todos;
+
+const selectVisibleTodos = createSelector(
```

Comment:
It's hopefully a temporary problem while we improve the state architecture, but should we mention `createDeepEqualSelector`?

Reply:
Based on a performance talk we had last week looking at Sentry reports, @Gudahtt mentioned that the work they are doing on controllers (Updated Controllers until all controllers extend Base controller V2.) they will use Immer under the hood (same as RTK), enabling us to only use shallow comparison which will be more performant than our deep comparison selectors.

Reply:
A current limitation in the extension is that the background sends all state to the frontend over a JSON-RPC API so it can't distinguish shallow reference changes.

But I'm actively working on a PR to resolve this, so happy for it not to be included here if it's been considered not best practice but a temporary workaround.